### PR TITLE
pytest for atf ; netlink support

### PR DESCRIPTION
**Pull request by AlexanderChernikov**

### CODE REVIEW

File: 2022q4/pytest-for-atf [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+link:https://reviews.freebsd.org/D31084[Initial review] URL: link:https://reviews.freebsd.org/D31084[https://reviews.freebsd.org/D31084] +
+link:https://cgit.freebsd.org/src/tree/tests/examples/test_examples.py[Test examples] URL: link:https://cgit.freebsd.org/src/tree/tests/examples/test_examples.py[https://cgit.freebsd.org/src/tree/tests/examples/test_examples.py]
+
+Contact: Alexander Chernikov <melifaro@FreeBSD.org>
+
+Native pytest support for atf(7) enhances the capabilities of the FreeBSD testing.
+
+Pytest simplifies the test writing by reducing the amount of boiler-plate code.
+It offers several advantages over the existing atf-c and atf-shell bindings.
+One of the most important ones is the test parametrization, which allows to improve coverage without writing nearly any code. The other is rich assert system, offering detailed errors description.
```
Comment (pauamma) [inline 2022q4/pytest-for-atf:13]: s/the test/test/

s/allows to improve/allows improving/

s/without writing nearly any code/with writing nearly no code/

s/is rich/is a rich/

File: 2022q4/pytest-for-atf [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+Python atf(7) support comes with a number of libraries that abstracts a number f commonly-used tasks. For example, running a test within a VNET jail with epair(4) requires adding a single line of code. Such helpers are especially handy in the networking area, where tests with complex VNET setups are not uncommon.
```
Comment (pauamma) [inline 2022q4/pytest-for-atf:14]: s/number f/number of/

File: 2022q4/pytest-for-atf [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+==== Current status
+Python support have been commited to HEAD. Currently, ~80 tests use the python framework and the number is rising. Example tests have beed commited to show the handling of the typical cases.
```
Comment (pauamma) [inline 2022q4/pytest-for-atf:17]: Needs blank line after header, IIRC.

Also, s/commited/committed/

File: 2022q4/pytest-for-atf [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+==== Next steps
+* Work on increasing the adoption of the framework
```
Comment (pauamma) [inline 2022q4/pytest-for-atf:20]: Blank line between those, as above.

**Comment by AlexanderChernikov** [comment]
Thank you! Will fix tomorrow morning if that’s okay

**Comment by pauamma** [comment]
That's fine.

File: 2022q4/netlink [hunk 5]
```diff
@@ -0,0 +1,31 @@
+=== Netlink on FreeBSD
+
+Links: +
+link:https://reviews.freebsd.org/D36002[Initial review] URL: link:https://reviews.freebsd.org/D36002[https://reviews.freebsd.org/D36002] +
+link:https://static.ipfw.ru/files/netlink.pdf[Netlink talk] URL: link:https://static.ipfw.ru/files/netlink.pdf[https://static.ipfw.ru/files/netlink.pdf]
+
+Contact: Alexander Chernikov <melifaro@FreeBSD.org>
+
+Netlink is a communication protocol defined in RFC 3549. It is async, TLV-based protocol, providing 1-1 and 1-many communications between kernel and userland. Netlink is currently used in Linux kernel to modify, read and subscribe for nearly all networking states. Interface state, addresses, routes, firewall, rules, fibs, etc, are controlled via Netlink.
```
Comment (grahamperrin) [inline 2022q4/netlink:9]: > It is async

Instead: 

> It is an async

Comment (grahamperrin) [reply 2022q4/netlink:9]: > used in Linux kernel

Instead: 

> used in the Linux kernel

File: 2022q4/netlink [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+==== Why Netlink is important for FreeBSD?
+POSIX defines API for base functions/system calls. There is no such standard for plethora of various protocol/device-level/subsystem-level ioctls. Each subsystem/driver invents its own protocol, handling format and compatibility.
```
Comment (grahamperrin) [inline 2022q4/netlink:12]: > standard for plethora of

Instead: 

> standard for the plethora of

File: 2022q4/netlink [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+Extendability is a notable problem in the networking control plane. For example, it is extremely hard to add properties to the routing socket messages without breaking compatibility.
+
+Netlink offers unification by providing standard communication layer and basic easily-extendable message formatting. It can serve as a "broker", automatically combining requested data from different sources in a single request (example: interface state dump). Netlink APIs allow to reduce the bar for application developers to support FreeBSD, while providing the desired extendability.
```
Comment (grahamperrin) [inline 2022q4/netlink:15]: > providing standard communication layer 

Instead: 

> providing a standard communication layer

File: 2022q4/netlink [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+==== Current status
+Netlink has been committed to HEAD. The code implements subset of the NETLINK_ROUTE subsystem and NETLINK_GENERIC framework.
+
+NETLINK_ROUTE supports add/delete/replace operations for routes, nexthops and link-level addresses. Partial support exists for the interface addresses and interfaces.
+
+Linuxolator support for Netlink has been committed to HEAD. It is possible to use unmodified ip with routes,nexthops, addresses and interfaces.
```
Comment (grahamperrin) [inline 2022q4/netlink:22]: > Linuxolator

Instead: 

> Linuxulator

Comment (grahamperrin) [reply 2022q4/netlink:22]: > routes,nexthops,

Instead: 

> routes, nexthops,

File: 2022q4/netlink [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+The simple userland library, snl(3), that provides convenient abstractions on the netlink socket, has been committed to HEAD.
+
+The first third-party software, bird, link:https://gitlab.nic.cz/labs/bird/-/commit/1e47b9f203aaaad0fb658d40a1670f1d0437f1f8[added] the experimental FreeBSD Netlink support.
```
Comment (grahamperrin) [inline 2022q4/netlink:26]: > the experimental

Instead: 

> experimental

**Review by grahamperrin — COMMENTED** [review]
I might make a PR for these suggested changes, for (AsciiDoc) line endings and blank lines, et cetera …

**Comment by grahamperrin** [comment]
> Thank you! Will fix tomorrow morning if that’s okay

Also, please rename the file to have an `.adoc` suffix. Thanks.

File: 2022q4/netlink.adoc [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+==== Current status
+
+Netlink has been committed to HEAD.
+The code implements a subset of the `NETLINK_ROUTE` subsystem and `NETLINK_GENERIC` framework.
+
+`NETLINK_ROUTE` supports add/delete/replace operations for routes, nexthops and link-level addresses.
+Partial support exists for the interface addresses and interfaces.
+
+Linuxulator support for Netlink has been committed to HEAD.
+It is possible to use unmodified ip with routes, nexthops, addresses and interfaces.
```
Comment (pauamma) [inline 2022q4/netlink.adoc:35]: "ip"? Do you mean "API" here?

File: 2022q4/netlink.adoc [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+
+The simple userland library, snl(3), that provides convenient abstractions on the netlink socket, has been committed to HEAD.
+
+The first third-party software, BIRD, link:https://gitlab.nic.cz/labs/bird/-/commit/1e47b9f203aaaad0fb658d40a1670f1d0437f1f8[added] experimental FreeBSD Netlink support.
```
Comment (pauamma) [inline 2022q4/netlink.adoc:39]:
```suggestion
The first third-party software, BIRD, link:https://gitlab.nic.cz/labs/bird/-/commit/1e47b9f203aaaad0fb658d40a1670f1d0437f1f8[added experimental FreeBSD Netlink support].
```

**Review by pauamma — CHANGES_REQUESTED** [review]

File: 2022q4/netlink.adoc (hunk 10 above)
Comment (AlexanderChernikov) [reply 2022q4/netlink.adoc:35]: Indeed it looks ambiguous. I meant "ip from iproute2".

File: 2022q4/netlink.adoc [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+==== Current status
+
+Netlink has been committed to HEAD.
+The code implements a subset of the `NETLINK_ROUTE` subsystem and `NETLINK_GENERIC` framework.
+
+`NETLINK_ROUTE` supports add/delete/replace operations for routes, nexthops and link-level addresses.
+Partial support exists for the interface addresses and interfaces.
+
+Linuxulator support for Netlink has been committed to HEAD.
+It is possible to use unmodified iproute2 ip binary with routes, nexthops, addresses and interfaces.
```
Comment (pauamma) [inline 2022q4/netlink.adoc:35]:
```suggestion
It is possible to use the unmodified ip from iproute2 with routes, nexthops, addresses and interfaces.
```

File: 2022q4/pytest-for-atf.adoc [hunk 13]
```diff
@@ -0,0 +1,23 @@
+=== Pytest support for the FreeBSD testing framework
+
+Links: +
+link:https://reviews.freebsd.org/D31084[Initial review] URL: link:https://reviews.freebsd.org/D31084[https://reviews.freebsd.org/D31084] +
+link:https://cgit.freebsd.org/src/tree/tests/examples/test_examples.py[Test examples] URL: link:https://cgit.freebsd.org/src/tree/tests/examples/test_examples.py[https://cgit.freebsd.org/src/tree/tests/examples/test_examples.py]
+
+Contact: Alexander Chernikov <melifaro@FreeBSD.org>
+
+Native pytest support for atf(7) enhances the capabilities of the FreeBSD testing.
```
Comment (pauamma) [inline 2022q4/pytest-for-atf.adoc:9]:
```suggestion
Native pytest support for atf(7) enhances the capabilities of the FreeBSD test suite.
```

Mostly a matter of taste.

File: 2022q4/pytest-for-atf.adoc [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+
+Pytest simplifies test writing by reducing the amount of boiler-plate code.
+It offers several advantages over the existing atf-c and atf-shell bindings.
+One of the most important ones is the test parametrization, which allows improving coverage with writing nearly no code. The other is a rich assert system, offering detailed errors description.
+Python atf(7) support comes with a number of libraries that abstracts a number of commonly-used tasks. For example, running a test within a VNET jail with epair(4) requires adding a single line of code. Such helpers are especially handy in the networking area, where tests with complex VNET setups are not uncommon.
+
+==== Current status
+
+Python support have been commited to HEAD. Currently, ~80 tests use the python framework and the number is rising. Example tests have beed commited to show the handling of the typical cases.
```
Comment (pauamma) [inline 2022q4/pytest-for-atf.adoc:18]:
```suggestion
Python support has been committed to HEAD. Currently, ~80 tests use the python framework and the number is rising. Example tests have been committed to show the handling of the typical cases.
```

**Review by pauamma — CHANGES_REQUESTED** [review]
Almost there.

**Review by pauamma — APPROVED** [review]

**Comment by AlexanderChernikov** [comment]
Thank you for the detailed review!

**Pull request merged**